### ISSUE TITLE
[BUG]: PCSX2 Crashes When Quitting PCSX2 in Fullscreen Mode

### ISSUE THREAD

**Issue by DanielBielak** [PCSX2/pcsx2#8704]
### Describe the Bug

PCSX2 crashes when quitting PCSX2 in fullscreen mode.

### Reproduction Steps

- Run Gran Turismo 4 with PCSX2
- Select Gran Turismo Mode
- Set screen to Fullscreen mode (I use keystroke Option+Return).
- Quit PCSX2 by pressing Command+Q.

### Expected Behavior

The quitting dialog box should appean with the option to quit the application and with the option to save state.

### PCSX2 Revision

v1.7.4398

### Operating System

Windows 11

### CPU

2.2 GHz Intel Core i7

### GPU

Intel Iris Pro 1536 MB

### GS Settings

Internal Resolution: 2x Native (~720p)

### Emulation Settings

Internal Resolution: 2x Native (~720p)
Clamping Mode: Full

### Logs & Dumps

[Dump.txt](https://github.com/PCSX2/pcsx2/files/11276546/Dump.txt)

**Comment by refractionpcsx2** [comment]
can you upload the .dmp file also? you should be able to put it in a zip file and upload it

**Comment by DanielBielak** [comment]
I uploaded a filed called dump.txt which contains the macOS error dump output. Where is the .dmp file located, or how do I create one?

**Comment by DigitalMajestic** [comment]
Regression from #8651

**Comment by Meerkov** [comment]
Possibly related: https://github.com/PCSX2/pcsx2/issues/8679 which is another freeze/crash related to exiting fullscreen.

**Comment by TellowKrinkle** [comment]
Full set of requirements to trigger this:
- Must be in render to main window
- Must have confirm before shutdown enabled

1. Emulator hits the code that's supposed to confirm on shutdown
2. ??? (Something fails to cancel quit request?)
3. App returns from exec and calls `EmuThread::stop();` https://github.com/PCSX2/pcsx2/blob/844f21344db286bae45e0e1e68a60173e9da6e50/pcsx2-qt/QtHost.cpp#L1836-L1840
4. Emulator shuts down without first going through `VMState::Stopping`
5. Shutdown happens in `processEvents`, and `Execute` is called immediately after, which tries to use the shutdown emulator's state and crashes https://github.com/PCSX2/pcsx2/blob/844f21344db286bae45e0e1e68a60173e9da6e50/pcsx2-qt/QtHost.cpp#L430-L433

@stenzek since you're a bit more familiar with qt, do you know what the things that might cause `app.exec()` to return are?

**Comment by TellowKrinkle** [comment]
Bisected to https://github.com/qt/qtbase/commit/a1ba8b726ec520487d2607de82b22487f4a1fd91

I know we do a lot of weird window manipulation for the render-to-main-window fullscreen case, @stenzek does anything in that commit's comment seem obvious how it would break sending a quit message while we're fullscreen during render-to-main-window?

**Comment by stenzek** [comment]
It's just another case I didn't think about during the whole shutdown flow. Dunno why people insist on using a hotkey to exit when the big picture UI is bound to escape by default... :/

> Also, what's the reason for the whole "temporarily leave fullscreen" thing here?

Exclusive fullscreen, and even without, drivers can scan the application buffers out directly.

**Comment by stenzek** [comment]
I guess we could ifdef it for Mac, if modal popups over a borderless fullscreen window are guaranteed to work there.

But I'll do a new Qt build for Windows and see if I can reproduce the issue first, because it works fine on 6.4.0.

**Comment by TellowKrinkle** [comment]
> Dunno why people insist on using a hotkey to exit when the big picture UI is bound to escape by default...

Cmd-Q is pretty much always the keyboard shortcut to quit on macOS, so I'm not surprised people use it

BTW I messed with it some more, and it seems the part that's breaking it is the fact that we're parenting the modal to a previously-not-visible window.  Maybe the setVisible doesn't process fast enough for some internal Qt thing to pick it up wrong?  Disabling the "hide main window when fullscreen and rendering to main" function fixes it, and it's also broken with fullscreen + render to separate + hide main

**Comment by stenzek** [comment]
Hmm, that's a bit of a difficult one then. I hide the main PCSX2 window when going fullscreen, because otherwise it's just a blank window which appears in the ALT-TAB list. Since we switch out of fullscreen to display the confirmation, that doesn't leave any visible windows which we could parent it to. 

So it should be parenting to the main window (which isn't visible at the time). Alternatively, we could parent to null, and hope it displays on top I guess.

**Comment by TellowKrinkle** [comment]
Were you able to repro on Windows?  If it's Mac specific, we can just leave the fullscreen window open on macOS (since it doesn't have exclusive fullscreen)

**Comment by stenzek** [comment]
It was fine in 6.4, I'm building 6.5 currently (process is more involved than on Linux/Mac sadly).

**Comment by stenzek** [comment]
I ended up redoing a large chunk of this logic.. mainly because of exclusive fullscreen being a pain in the arse.

The issue does happen on 6.5.0. So it must've changed in one of the 6.4 point releases. Regardless, I'm going out of my way to avoid tearing down all windows at *any* point in time. Turns out closing everything causes the main event loop to return, even if you open another window before returning to it.

For Macs, I left the popup display over the top of the fullscreen window. Hopefully this works, but I can't test it.

Windows was a bit messy. Exclusive fullscreen obviously needs you to switch back to windowed, but with driver direct scanout, what appeared behind the dialog was pretty inconsistent. Sometimes grey, sometimes black, sometimes the previous frame. So I made Windows always switch back to windowed.

Of course, fundamentally-broken-by-design screws this up too, because we can't change a window from windowed to fullscreen and vice versa in Wayland without things breaking. Switching back to windowed would mean tearing down the old window and recreating a new one, leaving us with the issue in the first paragraph. So for Wayland I just let it display over the top as well, hopefully it works... after rewriting 1k lines, I don't feel like rebooting into Linux to check tonight.

**Comment by TellowKrinkle** [comment]
> The issue does happen on 6.5.0. So it must've changed in one of the 6.4 point releases.

Yeah, the commit that broke it (https://github.com/qt/qtbase/commit/a1ba8b726ec520487d2607de82b22487f4a1fd91) was introduced in 6.4.1

> For Macs, I left the popup display over the top of the fullscreen window. Hopefully this works, but I can't test it.

Fairly sure that will work, I'll test it later though
